### PR TITLE
Fixed the Zero Division Error.

### PR DESCRIPTION
**Pull request by 3DSinghVFX**
Fixed the "ZeroDivisionError" showed by animation nodes in Blender2.83,
Screenshot from 2020-01-19 21-57-27

I have found that `@cython.cdivision(True)` is not written which gave the error,
https://github.com/JacquesLucke/animation_nodes/blob/2d7ddbc18a0fcf9056c191fea12a208e4afa0f36/animation_nodes/math/vector.pyx#L120-L124

### CODE REVIEW

**Comment by OmarEmaraDev** [comment]
`@cython.cdivision(True)` removes the zero division checks that Cython adds. We add this directive to make the division faster and avoid the overhead of zero division checks.

So you didn't really avoid the zero division here. Keep the directive but add an explicit check for zero division.

**Comment by OmarEmaraDev** [comment]
Which call to `angleVec3` resulted in the zero division?

**Comment by 3DSinghVFX** [comment]
> Which call to `angleVec3` resulted in the zero division?

Omar, I have tried with the same blend (and other files) but not able to regenerate the error.

**Comment by 3DSinghVFX** [comment]
> Which call to `angleVec3` resulted in the zero division?

Mesh From Spline node give the error when I trim the spline,
Screenshot from 2020-01-20 18-04-04

**Comment by OmarEmaraDev** [comment]
Ok.

**Comment by 3DSinghVFX** [comment]
When I commented this line `matrices = getMatricesAlongSpline(spline, amount, self.parameterType)` in the Mesh From Spline node then error disappeared.

**Comment by 3DSinghVFX** [comment]
> `@cython.cdivision(True)` removes the zero division checks that Cython adds. We add this directive to make the division faster and avoid the overhead of zero division checks.

Okay.

**Comment by OmarEmaraDev** [comment]
Will you continue working on this patch?

**Comment by 3DSinghVFX** [comment]
Omar, I have tried but I did not find anything related to `angleVec3` in the `matrices = getMatricesAlongSpline(spline, amount, self.parameterType)`.

**Comment by OmarEmaraDev** [comment]
`getMatricesAlongSpline` -> `ensureNormals` -> `calculateNormalsForTangents` -> `calcNextNormal` -> `angleVec3`.

In this patch, you just need to add an explicit check for zero.

**Comment by 3DSinghVFX** [comment]
Okay, Thanks :)

**Comment by OmarEmaraDev** [comment]
I actually meant in the `angleVec3` function.

**Comment by 3DSinghVFX** [comment]
Okay, I'll change that.

File: animation_nodes/math/vector.pyx [hunk 1]
```diff
@@ -118,10 +118,16 @@ cdef float distanceSquaredVec3(Vector3* a, Vector3* b):
...
     return a.x * b.x + a.y * b.y + a.z * b.z
 
+@cython.cdivision(True)
 cdef float angleVec3(Vector3 *a, Vector3 *b):
     cdef float dot = dotVec3(a, b)
-    cdef float val
-    val = dot / (lengthVec3(a) * lengthVec3(b))
+    cdef float val, lengthVecA, lengthVecB
+    lengthVecA = lengthVec3(a)
+    lengthVecB = lengthVec3(b)
```
Comment (OmarEmaraDev) [inline animation_nodes/math/vector.pyx]: I think you should just assign `lengthVec3(a) * lengthVec3(b)` to a single variable, then check for zero, returning zero directly.

Comment (3DSinghVFX) [reply animation_nodes/math/vector.pyx]: Okay.

Comment (3DSinghVFX) [reply animation_nodes/math/vector.pyx]: Changed.

**Pull request merged**